src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -0,0 +1,548 @@
+                                    return Ok(false);
+                                }
+                            } else {
+                                perm.latest_applied_was_write = Some(T::IS_WRITE);
+                            }
+                        } else {
+                            perm.latest_applied_was_write = None;
+                        }
+                        let old_perm = *perm;
+                        let new_perm = T::after_access(rel_pos, old_perm.data, protected)?;
+                        if old_perm.accessed {
+                            protector_allows_transition(global, node.tag, old_perm.data, new_perm)?
+                        }
+                        perm.data = new_perm;
+                        perm.accessed |= !rel_pos.is_foreign();
+                        Ok(true)
+                    },
+                    |args: ErrHandlerArgs<TransitionError>| -> InterpErrorInfo<'tcx> {
+                        let ErrHandlerArgs {
+                            error: error_kind,
+                            rel_pos: rel_pos_of_access,
+                            faulty_tag,
+                        } = args;
+                        TbError {
+                            faulty_tag,
+                            access_kind: T::NAME,
+                            error_kind,
+                            initial_tag: tag,
+                            rel_pos_of_access,
+                        }
+                        .build()
+                    },
+                )?;
+        }
+        Ok(())
+    }
+}
+
+/// Integration with the BorTag garbage collector
+impl Tree {
+    pub fn remove_unreachable_tags(&mut self, live_tags: &FxHashSet<BorTag>) {
+        assert!(self.keep_only_needed(self.root, live_tags)); // root can't be removed
+    }
+
+    /// Traversal function: returns true iff the tag is necessary. A tag is necessary
+    /// if it is either live or it has necessary children.
+    fn keep_only_needed(&mut self, idx: UniIndex, live: &FxHashSet<BorTag>) -> bool {
+        let node = self.nodes.get(idx).unwrap();
+        let children: SmallVec<_> = node
+            .children
+            .clone()
+            .into_iter()
+            .filter(|child| self.keep_only_needed(*child, live))
+            .collect();
+        let no_children = children.is_empty();
+        let node = self.nodes.get_mut(idx).unwrap();
+        node.children = children;
+        // All of the children are unreachable
+        if !live.contains(&node.tag) && no_children {
+            // Important: make sure to remove _all_ of them because we're going to forget the index
+            // afterwards.
+            let tag = node.tag;
+            self.nodes.remove(idx);
+            for perms in self.rperms.iter_mut_all() {
+                perms.remove(idx);
+            }
+            self.tag_mapping.forget(&tag);
+            false
+        } else {
+            true
+        }
+    }
+}
```

Comment:
Might be worth noting somewhere that this leave cleanup potential on the table: if a dead tag is in the middle of the tree, we might still be able to "compact" the tree and re-assign its children to directly link to its parent. It's subtle though since we must make sure that this doesn't lead to any UB being missed.